rust/frontend/src/ndarray.rs

```diff
@@ -0,0 +1,359 @@
+    /// ## Example
+    ///
+    /// ```
+    /// let mut shape = vec![2];
+    /// let mut data = vec![1f32, 2];
+    /// let ctx = TVMContext::gpu(0);
+    /// let mut ndarray = empty(&mut shape, ctx, TVMType::from("int"));
+    /// ndarray.copy_from_buffer(&mut data);
+    /// ```
+    pub fn copy_from_buffer<T: Num32>(&mut self, data: &mut [T]) {
+        check_call!(ts::TVMArrayCopyFromBytes(
+            self.handle,
+            data.as_ptr() as *mut _,
+            data.len() * mem::size_of::<T>()
+        ));
+    }
+
+    /// Copies the NDArray to another target NDArray.
+    pub fn copy_to_ndarray(&self, target: NDArray) -> Result<NDArray> {
+        if self.dtype() != target.dtype() {
+            return Err(Error::TypeMismatch {
+                expected: self.dtype().to_string(),
+                found: target.dtype().to_string(),
+            });
+        }
+        check_call!(ts::TVMArrayCopyFromTo(
+            self.handle,
+            target.handle,
+            ptr::null_mut() as ts::TVMStreamHandle
+        ));
+        Ok(target)
+    }
+
+    /// Copies the NDArray to a target context.
+    pub fn copy_to_ctx(&self, target: &TVMContext) -> Result<NDArray> {
+        let tmp = empty(&mut self.shape().unwrap(), target.clone(), self.dtype());
+        let copy = self.copy_to_ndarray(tmp)?;
+        Ok(copy)
+    }
+
+    /// Converts a Rust's ndarray to TVM NDArray.
+    pub fn from_rust_ndarray<T: Num32 + Copy>(
+        rnd: &ArrayD<T>,
+        ctx: TVMContext,
+        dtype: TVMType,
+    ) -> Result<Self> {
+        let mut shape = rnd.shape().to_vec();
+        let mut nd = empty(&mut shape, ctx, dtype);
+        let mut vec = Array::from_iter(rnd.iter())
+            .iter()
+            .map(|e| **e)
+            .collect::<Vec<T>>();
+        nd.copy_from_buffer(vec.as_mut_slice());
+        Ok(nd)
+    }
+}
+
+/// Creates an empty NDArray given shape, context and dtype.
+pub fn empty(shape: &mut Vec<usize>, ctx: TVMContext, dtype: TVMType) -> NDArray {
+    let mut handle = ptr::null_mut() as ts::TVMArrayHandle;
+    check_call!(ts::TVMArrayAlloc(
+        shape.as_ptr() as *const i64,
+        shape.len() as c_int,
+        dtype.inner.code as c_int,
+        dtype.inner.bits as c_int,
+        dtype.inner.lanes as c_int,
+        ctx.device_type.0 as c_int,
+        ctx.device_id as c_int,
+        &mut handle as *mut _,
+    ));
+    NDArray::new(handle, false)
+}
+
+macro_rules! impl_from_ndarray_rustndarray {
+    ($type:ty, $type_name:tt) => {
+        impl<'a> TryFrom<&'a NDArray> for ArrayD<$type> {
+            type Error = Error;
+            fn try_from(nd: &NDArray) -> Result<ArrayD<$type>> {
+                if nd.shape().is_none() {
+                    return Err(Error::EmptyArray);
+                }
+                assert_eq!(nd.dtype(), TVMType::from($type_name), "Type mismatch");
+                Ok(
+                    Array::from_shape_vec(nd.shape().unwrap().clone(), nd.to_vec::<$type>()?)
```

Comment:
no unwrap

Reply:
`nd.shape().unwrap()` has been capture in the if clause earlier. For the last `unwrap` inside `Ok`, it'll be resolved if I switch to `error_chain` and add `ShapeError` or at I'll need to add it to `custom_error`. Let me see.

Reply:
Well, `error_chain` resolves it nicely.